### PR TITLE
[IOTDB-5117] Introduce MemoryDistributionCalculator for FragmentInstance

### PR DESCRIPTION
**Pull request by lancelly**
details could be found at: https://apache-iotdb.feishu.cn/docx/O5XCdvEHPoDdtNxFHlXcV6xlnoh

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/SinkHandle.java [hunk 1]
```diff
@@ -302,6 +337,15 @@ public TFragmentInstanceId getLocalFragmentInstanceId() {
     return localFragmentInstanceId;
   }
 
+  public long getMaxBytesCanReserve() {
+    return maxBytesCanReserve;
+  }
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/SinkHandle.java:342]:
```suggestion
```

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/SourceHandle.java [hunk 2]
```diff
@@ -337,6 +374,15 @@ public long getBufferRetainedSizeInBytes() {
     return bufferRetainedSizeInBytes;
   }
 
+  public long getMaxBytesCanReserve() {
+    return maxBytesCanReserve;
+  }
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/SourceHandle.java:379]:
```suggestion
```

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java [hunk 3]
```diff
@@ -0,0 +1,615 @@
...
+      exchangeMap
+          .computeIfAbsent(node.getDownStreamPlanNodeId(), x -> new ArrayList<>())
+          .add(node.getDownStreamPlanNodeId());
+    }
+    return null;
+  }
+
+  @Override
+  public Void visitSeriesScan(SeriesScanNode node, MemoryDistributionContext context) {
+    processConsumeAllChildrenAtTheSameTime(node);
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java:192]:
```suggestion
```

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java [hunk 4]
```diff
@@ -0,0 +1,615 @@
...
+    return null;
+  }
+
+  @Override
+  public Void visitSeriesAggregationScan(
+      SeriesAggregationScanNode node, MemoryDistributionContext context) {
+    processConsumeAllChildrenAtTheSameTime(node);
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java:199]:
```suggestion
```

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java [hunk 5]
```diff
@@ -0,0 +1,615 @@
...
+    return null;
+  }
+
+  @Override
+  public Void visitAlignedSeriesScan(
+      AlignedSeriesScanNode node, MemoryDistributionContext context) {
+    processConsumeAllChildrenAtTheSameTime(node);
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java:206]:
```suggestion
```

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java [hunk 6]
```diff
@@ -0,0 +1,615 @@
...
+    return null;
+  }
+
+  @Override
+  public Void visitAlignedSeriesAggregationScan(
+      AlignedSeriesAggregationScanNode node, MemoryDistributionContext context) {
+    processConsumeAllChildrenAtTheSameTime(node);
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java:213]:
```suggestion
```

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java [hunk 7]
```diff
@@ -0,0 +1,615 @@
...
+  @Override
+  public Void visitAlignedLastQueryScan(
+      AlignedLastQueryScanNode node, MemoryDistributionContext context) {
+    processConsumeAllChildrenAtTheSameTime(node);
+    return null;
+  }
+
+  @Override
+  public Void visitLastQuery(LastQueryNode node, MemoryDistributionContext context) {
+    processConsumeAllChildrenAtTheSameTime(node);
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java:318]: actually, `LastQueryOperator` and `LastQuerySortOperator` should call `processConsumeChildrenOneByOneNode`.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java [hunk 8]
```diff
@@ -0,0 +1,615 @@
...
+
+  @Override
+  public Void visitLastQueryMerge(LastQueryMergeNode node, MemoryDistributionContext context) {
+    processConsumeAllChildrenAtTheSameTime(node);
+    return null;
+  }
+
+  @Override
+  public Void visitLastQueryCollect(LastQueryCollectNode node, MemoryDistributionContext context) {
+    processConsumeAllChildrenAtTheSameTime(node);
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java:330]:
```suggestion
    processConsumeChildrenOneByOneNode(node);
```

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java (hunk 7 above)
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java:318]:
```suggestion
    processConsumeChildrenOneByOneNode(node);
```

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java [hunk 9]
```diff
@@ -0,0 +1,615 @@
...
+  @Override
+  public Void visitSchemaFetchScan(SchemaFetchScanNode node, MemoryDistributionContext context) {
+    processConsumeAllChildrenAtTheSameTime(node);
+    return null;
+  }
+
+  @Override
+  public Void visitCreateAlignedTimeSeries(
+      CreateAlignedTimeSeriesNode node, MemoryDistributionContext context) {
+    processConsumeAllChildrenAtTheSameTime(node);
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java:431]: all the write operation should throw exception

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/SinkHandle.java (hunk 1 above)
Comment (lancelly) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/SinkHandle.java]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/SourceHandle.java (hunk 2 above)
Comment (lancelly) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/SourceHandle.java]: done

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java (hunk 3 above)
Comment (lancelly) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java]: done

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java (hunk 4 above)
Comment (lancelly) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java]: done

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java (hunk 5 above)
Comment (lancelly) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java]: done

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java (hunk 6 above)
Comment (lancelly) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java]: done

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java (hunk 7 above)
Comment (lancelly) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java]: done

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java (hunk 8 above)
Comment (lancelly) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java]: done

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java (hunk 9 above)
Comment (lancelly) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/MemoryDistributionCalculator.java]: done

**Review by JackieTien97 — APPROVED** [review]

**Pull request merged**